<H1>SQL test for SumUp Analyst</H1>

by: Diego Rivera

<H3>Objective:</H3>

The main objective of this document is to respond to the challenge proposed for the SumUp Analyst process, following the recommendations and instructions, within the proposed scope.

The work carried out is summarized in:

<ul>
  <li>Data capture and storage</li>
  <li>Data review and cleaning</li>
  <li>Resolution of the questions raised</li>
  <ul>
    <li>Data Inspection</li>
    <li>Selection of data to use</li>
    <li>Implementation of functions for processing</li>
    <li>Response declaration </li>
  </ul>
</ul>

<i><b>Note:</b></i> Given the characteristics of the <i>CSV</i> files that made up the data set delivered to obtain the data to be analyzed, I opted for the exclusive use of Python3 for the resolution of the test, since I do not have a server available that supports the <i >insert</i> of more than 780 thousand records for a table (a second table required the insert of more than 600 thousand records) and at the moment I only had my MAMP localhost available.

That is why I worked on <i>Jupyter Notebook (Anaconda)</i>, an interactive work environment that allows me to develop code in Python dynamically, while integrating both blocks of code and text (markdown) in the same document, to prepare this report in a simple and effective way as it is easier to detect errors (by doing it by blocks) avoiding the code processing crash as occurs in other environments such as traditional IDEs.

=====================================================================================================================

Importing the main libraries to facilitate the manipulation of data in the development of the test:

* <i><b>Pandas</b></i> library will be used mainly for the management and analysis of data structures such as lists, series and data frames.

* <i><b>NumPy</b></i> library will be used mainly in the numerical - mathematical calculation and data analysis, of the large volume of data with which it will work.

In [13]:
import pandas as pd
import numpy as np

The capture and storage of the data to be processed for the development of the test was carried out through Pandas' own function, <i>pd.read_csv()</i>, where the argument of the function will be a variable that will have as a value the directory path of each CSV file delivered by SumUp. This function will read each of the CSV files to store the content of the data in variables named appropriately, according to their content, to facilitate the identification and use according to the need of the moment. The format of these data structures will be of the DataFrame type, a Pandas-specific format to facilitate the correct handling of this volume of data.

In [259]:
bank_path = "data_bank_accounts.csv"
bank_accounts = pd.read_csv(bank_path)

In [260]:
bank_accounts

id bank_name     account_type
0      6834904.0    BANK 1    VISTA_ACCOUNT
1      4669208.0    BANK 1    VISTA_ACCOUNT
2      6945969.0    BANK 1  SAVINGS_ACCOUNT
3      3688345.0    BANK 1    VISTA_ACCOUNT
4      9271171.0    BANK 1    VISTA_ACCOUNT
...          ...       ...              ...
40145  9037293.0    BANK 1    VISTA_ACCOUNT
40146  8682059.0    BANK 1  SAVINGS_ACCOUNT
40147  9081287.0    BANK 2  CURRENT_ACCOUNT
40148  9301512.0    BANK 1    VISTA_ACCOUNT
40149  9273893.0    BANK 1    VISTA_ACCOUNT

[40150 rows x 3 columns]

In [261]:
cards_path = "data_cards.csv"
cards = pd.read_csv(cards_path)

In [17]:
cards

id card_type  card_brand
0       455163604     DEBIT        VISA
1       418035837     DEBIT        VISA
2       442446705     DEBIT        VISA
3       407122304     DEBIT  MASTERCARD
4       463103208     DEBIT        VISA
...           ...       ...         ...
609806  280062012     DEBIT        VISA
609807  448901621     DEBIT        VISA
609808  255625057     DEBIT        VISA
609809  328303062     DEBIT        VISA
609810  300043014     DEBIT        VISA

[609811 rows x 3 columns]

In [262]:
merchant_path = "data_merchants.csv"
merchants = pd.read_csv(merchant_path)

In [263]:
merchants

id     merchant_category  bank_account_id  comuna_id
0       5137058     Food Truck / Cart        6834904.0      493.0
1       4457960     Food Truck / Cart        4669208.0      693.0
2       8669992        Food / Grocery        6945969.0      544.0
3       5127389          Other Retail        3688345.0      700.0
4      10842385        Food / Grocery        9271171.0      493.0
...         ...                   ...              ...        ...
40245  10432888               Apparel        9037293.0      759.0
40246  10364267       Beauty / Barber        8682059.0      559.0
40247  10609760  Fast Food Restaurant        9081287.0      525.0
40248  10873272        Imported Goods        9301512.0      733.0
40249   9586213        Food / Grocery        9273893.0      546.0

[40250 rows x 4 columns]

In [264]:
trans_path = "data_transactions.csv"
transactions = pd.read_csv(trans_path)

In [265]:
transactions

id   date  merchant_id    card_id card_entry_mode   amount  \
0       1605032286  DAY 1      5137058  455163604          MODE 1  24000.0   
1       1605032299  DAY 1      4457960  418035837          MODE 2   4500.0   
2       1605032302  DAY 1      8669992  442446705          MODE 2   3200.0   
3       1605032254  DAY 1      4457960  418035837          MODE 1   4500.0   
4       1605032260  DAY 1      5127389  407122304          MODE 1  12000.0   
...            ...    ...          ...        ...             ...      ...   
784021  1616153350  DAY 3      6857379  328303062          MODE 2   4200.0   
784022  1616153371  DAY 3      9286781  300043014          MODE 1   2700.0   
784023  1616153397  DAY 3      5907093  239240589          MODE 3   6000.0   
784024  1616153405  DAY 3      1428282  426047641          MODE 1   7500.0   
784025  1616153378  DAY 3      4708365  426731435          MODE 1   5000.0   

            result  
0       SUCCESSFUL  
1       SUCCESSFUL  
2       SUCCESSFUL  
3           FAILED  
4       SUCCESSFUL  
...            ...  
784021      FAILED  
784022  SUCCESSFUL  
784023  SUCCESSFUL  
784024  SUCCESSFUL  
784025  SUCCESSFUL  

[784026 rows x 7 columns]

In [26]:
# checking the minimun value into amount variable
transactions['amount'].min()

200.0

After having all the data loaded in their respective DataFrames, I will make a brief review of the content of each variable with which I will work. 

Given the objective of the case, I will only review the presence of missing or null data and I will not deepen a review of descriptive statistics as in the case of outliers, since the objective is not to create any predictive or prescriptive model, in addition to being treated in the majority of descriptive variables and identification id's and the only quantitative value that could suffer distortion, is the amount in CLP of the transactions carried out. 

According to what was consulted on the internet, the minimum possible to pay with SumUp in Chile corresponds to <i>CLP $200</i>, which is precisely the minimum value registered in the respective DataFrame. It will be assumed not to have a maximum amount for this value, since it will depend both on the business where the transaction was made and the product or service paid and the respective customer who paid.

In [29]:
# A missing id. The record will be deleted so that it does not affect the possible results to be obtained. 
# Said removal will not affect the display of the 'bank_accounts' data as this is 1 record to delete out of 40,150
# records.
bank_accounts.isnull().sum()

id              1
bank_name       0
account_type    0
dtype: int64

In [30]:
# A missing card_type and card_brand. The records will be deleted so that it does not affect the possible results 
# to be obtained. Said removal will not affect the display of the 'cards' data since it is a maximum of 2 records
# to be deleted out of 609,811 records.
cards.isnull().sum()

id            0
card_type     1
card_brand    1
dtype: int64

In [31]:
# 101 missing bank_account_id and 17 comuna_id. The records will be deleted so that it does not affect the possible
# results to be obtained. Said removal will not affect the sample of the 'merchants' data since it involves a maximum
# of 118 records to be deleted out of 40,250 records (0.29% of the records).
merchants.isnull().sum()

id                     0
merchant_category      0
bank_account_id      101
comuna_id             17
dtype: int64

In [32]:
# 2056 card_entry_mode missing. The records will be deleted so that it does not affect the possible results to be 
# obtained. Said removal will not affect the display of the 'transactions' data since there are 2056 records to be
# deleted out of 784,026 records (0.26% of the records).
transactions.isnull().sum()

id                    0
date                  0
merchant_id           0
card_id               0
card_entry_mode    2056
amount                0
result                0
dtype: int64

In [38]:
# Dropping all null values for all DataFrames
bank_accounts = bank_accounts.dropna()
cards = cards.dropna()
merchants = merchants.dropna()
transactions = transactions.dropna()

In [39]:
# Checking for the correct dropping for all null values in the DataFrame 'bank_acount'
bank_accounts.isnull().sum()

id              0
bank_name       0
account_type    0
dtype: int64

In [40]:
# Checking for the correct dropping for all null values in the DataFrame 'cards'
cards.isnull().sum()

id            0
card_type     0
card_brand    0
dtype: int64

In [41]:
# Checking for the correct dropping for all null values in the DataFrame 'merchants'
merchants.isnull().sum()

id                   0
merchant_category    0
bank_account_id      0
comuna_id            0
dtype: int64

In [42]:
# Checking for the correct dropping for all null values in the DataFrame 'transactions'
transactions.isnull().sum()

id                 0
date               0
merchant_id        0
card_id            0
card_entry_mode    0
amount             0
result             0
dtype: int64

=====================================================================================================================

<H3>Questions</H3>

=====================================================================================================================


<b>1.</b> Create a query that allows you to see daily sales for each merchant.

In [51]:
# Merging the DataFrame se transactions and merchants, similar logic of the 'join' clause in SQL.
# 'transactions' would be the left DataFrame (same logic of left table y SQL) and 'merchant' the right one.
# Assigning the DataFrame to get in the variable 'answer1_prev'
answer1_prev = pd.merge(transactions[transactions.result == 'SUCCESSFUL'], merchants, how = 'inner',
                   left_on = 'merchant_id', right_on = 'id')

I consider that to better answer this question, I will only work with the data corresponding to 'Successful' transactions given by the restriction <i>'transactions[transactions.result == 'SUCCESSFUL']'</i> within the selection of the DataFrame transactions. The latter has the same scope as an SQL <i>'WHERE'</i> clause (WHERE t.result = 'SUCCESSFUL').

There is also the option to omit this restriction and consider all the transactions made, both <i>'Successful'</i> and <i>'Failed'</i>.

In [55]:
# Preview the result of both DataFrame columns together
answer1_prev

id_x   date  merchant_id    card_id card_entry_mode   amount  \
0       1605032286  DAY 1      5137058  455163604          MODE 1  24000.0   
1       1605033290  DAY 1      5137058  443503288          MODE 1  16200.0   
2       1605034956  DAY 1      5137058  312533087          MODE 1   3700.0   
3       1608207903  DAY 1      5137058  443348591          MODE 1  15000.0   
4       1608211681  DAY 1      5137058  406650881          MODE 1  42000.0   
...            ...    ...          ...        ...             ...      ...   
746665  1616149794  DAY 3     10432888  379590154          MODE 1   6000.0   
746666  1616150907  DAY 3     10364267  440949057          MODE 2  15000.0   
746667  1616151077  DAY 3     10609760  205956465          MODE 2   5000.0   
746668  1616152016  DAY 3     10873272  326715196          MODE 1   1000.0   
746669  1616152280  DAY 3      9586213  309161578          MODE 3  24000.0   

            result      id_y     merchant_category  bank_account_id  comuna_id  
0       SUCCESSFUL   5137058     Food Truck / Cart        6834904.0      493.0  
1       SUCCESSFUL   5137058     Food Truck / Cart        6834904.0      493.0  
2       SUCCESSFUL   5137058     Food Truck / Cart        6834904.0      493.0  
3       SUCCESSFUL   5137058     Food Truck / Cart        6834904.0      493.0  
4       SUCCESSFUL   5137058     Food Truck / Cart        6834904.0      493.0  
...            ...       ...                   ...              ...        ...  
746665  SUCCESSFUL  10432888               Apparel        9037293.0      759.0  
746666  SUCCESSFUL  10364267       Beauty / Barber        8682059.0      559.0  
746667  SUCCESSFUL  10609760  Fast Food Restaurant        9081287.0      525.0  
746668  SUCCESSFUL  10873272        Imported Goods        9301512.0      733.0  
746669  SUCCESSFUL   9586213        Food / Grocery        9273893.0      546.0  

[746670 rows x 11 columns]

In [56]:
# Selecting the proper columns in 'answer1_prev' to repond questions number 1
# Assigning this second DataFrame to get in the variable 'answer1'
answer1 = answer1_prev.loc[:,['merchant_id','merchant_category', 'date','amount']]

Given the data provided, I consider that the best view to illustrate a query that allows me to see daily sales for each merchant is the one contained in the variable <i>'answer1'</i>.

In [57]:
# Preview on the final answer for question 1
answer1

merchant_id     merchant_category   date   amount
0           5137058     Food Truck / Cart  DAY 1  24000.0
1           5137058     Food Truck / Cart  DAY 1  16200.0
2           5137058     Food Truck / Cart  DAY 1   3700.0
3           5137058     Food Truck / Cart  DAY 1  15000.0
4           5137058     Food Truck / Cart  DAY 1  42000.0
...             ...                   ...    ...      ...
746665     10432888               Apparel  DAY 3   6000.0
746666     10364267       Beauty / Barber  DAY 3  15000.0
746667     10609760  Fast Food Restaurant  DAY 3   5000.0
746668     10873272        Imported Goods  DAY 3   1000.0
746669      9586213        Food / Grocery  DAY 3  24000.0

[746670 rows x 4 columns]

=====================================================================================================================
<b>2.</b> What is the average sale amount by card type (credit/debit card)?

        a. How could you explain the difference between both card types?

In [58]:
# This time I will immediately select the columns to work with, both from the 'transactions' DataFrame
# as of 'cards', assigning them in the variables 'trans_prev2' and 'card_prev2', respectively. The 'loc' function
# has the same scope of the SQL clause 'SELECT'
trans_prev2 = transactions.loc[:,['card_id','amount','result']]
card_prev2 = cards.loc[:,['card_type', 'id']]

In [59]:
# Preview the result of 'trans_prev2' DataFrame, with the selected 'transactions' columns
trans_prev2

card_id   amount      result
0       455163604  24000.0  SUCCESSFUL
1       418035837   4500.0  SUCCESSFUL
2       442446705   3200.0  SUCCESSFUL
3       418035837   4500.0      FAILED
4       407122304  12000.0  SUCCESSFUL
...           ...      ...         ...
784021  328303062   4200.0      FAILED
784022  300043014   2700.0  SUCCESSFUL
784023  239240589   6000.0  SUCCESSFUL
784024  426047641   7500.0  SUCCESSFUL
784025  426731435   5000.0  SUCCESSFUL

[781970 rows x 3 columns]

In [60]:
# Preview the result of 'card_prev2' DataFrame, with the selected 'cards' columns
card_prev2

card_type         id
0          DEBIT  455163604
1          DEBIT  418035837
2          DEBIT  442446705
3          DEBIT  407122304
4          DEBIT  463103208
...          ...        ...
609806     DEBIT  280062012
609807     DEBIT  448901621
609808     DEBIT  255625057
609809     DEBIT  328303062
609810     DEBIT  300043014

[609810 rows x 2 columns]

In [65]:
# Merging 'trans_prev2' (left DataFrame) and 'card_prev2' (right DataFrame).
# To give this answer a better scope, it was again filtered by 'Successful' transactions.
# Assigning the DataFrame to get in the variable 'answer2_prev'.
answer2_prev = pd.merge(trans_prev2[trans_prev2.result == 'SUCCESSFUL'], card_prev2, how = 'inner',
                   left_on = 'card_id', right_on = 'id')

In [83]:
# Preview on 'answer2_prev' DataFrame, checking the proper columns selected and the data volume to prepare the
# final answer for question 2
answer2_prev

card_id   amount      result card_type         id
0       455163604  24000.0  SUCCESSFUL     DEBIT  455163604
1       418035837   4500.0  SUCCESSFUL     DEBIT  418035837
2       442446705   3200.0  SUCCESSFUL     DEBIT  442446705
3       407122304  12000.0  SUCCESSFUL     DEBIT  407122304
4       463103208  12000.0  SUCCESSFUL     DEBIT  463103208
...           ...      ...         ...       ...        ...
747517  280062012   5300.0  SUCCESSFUL     DEBIT  280062012
747518  448901621   6570.0  SUCCESSFUL     DEBIT  448901621
747519  255625057   2000.0  SUCCESSFUL     DEBIT  255625057
747520  300043014   2700.0  SUCCESSFUL     DEBIT  300043014
747521  239240589   6000.0  SUCCESSFUL     DEBIT  239240589

[747522 rows x 5 columns]

In [93]:
# Doing a grouping by 'card_type' and calculating the average 'amount' of the transactions and assigning the
# result in variable 'answer2_prev2'. I must emphasize that given the conditions of the filter and calculation
# the result will be a data structure type of 'Series'
answer2_prev2 = answer2_prev.groupby(['card_type'])['amount'].mean()

In [72]:
# data type structure
type(answer2_prev2)

pandas.core.series.Series

In [96]:
# Preview on 'answer2_prev2' Series. The result is in the Series preview, however being a data structure of type
# Series, it is possible to better illustrate of the result with an iterative FOR loop.
answer2_prev2

card_type
CREDIT    18205.845732
DEBIT      9932.567421
Name: amount, dtype: float64

In [99]:
# Question 2 answer
print('The average of the sale amount by card type is:')
for j in range(0,answer2_prev2.size):
    print('- ',answer2_prev2.index[j],': CLP ${:,.2f}'.format(answer2_prev2[j]))

The average of the sale amount by card type is:
-  CREDIT : CLP $18,205.85
-  DEBIT : CLP $9,932.57


That the average sales per card type is almost double in CREDIT card, <i>CLP 18,205.85</i>, versus <i>CLP 9,932.57</i> of the average sales of DEBIT card, could be explained by several factors such as:

The lack of circulating liquidity (or cash) in the economy, beyond the lack of 'coins' and 'bills' that people have in their pockets and merchants in their petty cash, but rather the liquidity available in people's checking accounts. Given the complicated events of recent years with the social outbreak of 2019 and the last years of the pandemic, the amount of circulating liquidity in our economy has decreased considerably, to the point that the president of the Central Bank of Chile asked people to retail markets will try to pay with coins, because these no longer had to give change back to people.

The withdrawals from the AFPs, although they meant purchasing power for many people, with precisely cash, lasted very little and in the end, it translated into two-digit inflation for Chile. Which has not been seen in many years.

The rising cost of living also affects the fact that people can no longer access certain goods and services by paying 'cash' or debit, and the use of credit and instalments is necessary to maintain the lifestyle that people want. Beyond the loyalty plans of the banks and the credit cards themselves, I think that in general there is a better understanding of the use of credit cards by people, of the use of the instalments that can be paid with or without interest. and to make payments on time, keeping in mind the billing dates.


<i><b>Source:</b></i> https://www.bcentral.cl/contenido/-/detalle/presidente-del-banco-central-promueve-uso-de-monedas-para-payments-en-efectivo

=====================================================================================================================
<b>3.</b> What is the merchant category that made the most succesful transactions during day 1?

In [106]:
# Again I will select the columns to analyze at the beginning, assigning the values of 'transactions' DataFrame in
# the 'trans_prev3' variables and the 'merchants' DataFrame in the 'merch_prev3' variable.
trans_prev3 = transactions.loc[:,['merchant_id','result','date']]
merch_prev3 = merchants.loc[:,['merchant_category', 'id']]

In [107]:
# Preview the result of 'trans_prev3' DataFrame, with the selected 'transactions' columns
trans_prev3

merchant_id      result   date
0           5137058  SUCCESSFUL  DAY 1
1           4457960  SUCCESSFUL  DAY 1
2           8669992  SUCCESSFUL  DAY 1
3           4457960      FAILED  DAY 1
4           5127389  SUCCESSFUL  DAY 1
...             ...         ...    ...
784021      6857379      FAILED  DAY 3
784022      9286781  SUCCESSFUL  DAY 3
784023      5907093  SUCCESSFUL  DAY 3
784024      1428282  SUCCESSFUL  DAY 3
784025      4708365  SUCCESSFUL  DAY 3

[781970 rows x 3 columns]

In [108]:
# Preview the result of 'merch_prev3' DataFrame, with the selected 'merchants' columns
merch_prev3

merchant_category        id
0         Food Truck / Cart   5137058
1         Food Truck / Cart   4457960
2            Food / Grocery   8669992
3              Other Retail   5127389
4            Food / Grocery  10842385
...                     ...       ...
40245               Apparel  10432888
40246       Beauty / Barber  10364267
40247  Fast Food Restaurant  10609760
40248        Imported Goods  10873272
40249        Food / Grocery   9586213

[40132 rows x 2 columns]

In [163]:
# As I did before, merging 'trans_prev3' (left DataFrame) and 'merch_prev3' (right DataFrame).
# This time filtering by 'Successful' transactions is required.
# Assigning the DataFrame to get in the variable 'answer3_prev'.
answer3_prev = pd.merge(trans_prev3[trans_prev3.result == 'SUCCESSFUL'], merch_prev3, how = 'inner',
                   left_on = 'merchant_id', right_on = 'id')

In [164]:
# Preview on 'answer3_prev' DataFrame, checking the proper columns selected and the data volume to move forward with
# this question.
answer3_prev

merchant_id      result   date     merchant_category        id
0           5137058  SUCCESSFUL  DAY 1     Food Truck / Cart   5137058
1           5137058  SUCCESSFUL  DAY 1     Food Truck / Cart   5137058
2           5137058  SUCCESSFUL  DAY 1     Food Truck / Cart   5137058
3           5137058  SUCCESSFUL  DAY 1     Food Truck / Cart   5137058
4           5137058  SUCCESSFUL  DAY 1     Food Truck / Cart   5137058
...             ...         ...    ...                   ...       ...
746665     10432888  SUCCESSFUL  DAY 3               Apparel  10432888
746666     10364267  SUCCESSFUL  DAY 3       Beauty / Barber  10364267
746667     10609760  SUCCESSFUL  DAY 3  Fast Food Restaurant  10609760
746668     10873272  SUCCESSFUL  DAY 3        Imported Goods  10873272
746669      9586213  SUCCESSFUL  DAY 3        Food / Grocery   9586213

[746670 rows x 5 columns]

In [115]:
# Filtering 'answer3_prev' variable with only transactions made on DAY 1.
answer3_prev = answer3_prev[answer3_prev['date'] == 'DAY 1']

In [116]:
# Preview on 'answer3_prev' DataFrame, checking the proper columns selected and the data volume to move forward with
# this question.
answer3_prev

merchant_id      result   date     merchant_category        id
0           5137058  SUCCESSFUL  DAY 1     Food Truck / Cart   5137058
1           5137058  SUCCESSFUL  DAY 1     Food Truck / Cart   5137058
2           5137058  SUCCESSFUL  DAY 1     Food Truck / Cart   5137058
3           5137058  SUCCESSFUL  DAY 1     Food Truck / Cart   5137058
4           5137058  SUCCESSFUL  DAY 1     Food Truck / Cart   5137058
...             ...         ...    ...                   ...       ...
686311      8665938  SUCCESSFUL  DAY 1  Fast Food Restaurant   8665938
686315     10294916  SUCCESSFUL  DAY 1        Food / Grocery  10294916
686316     10542065  SUCCESSFUL  DAY 1          Other Retail  10542065
686324     10017961  SUCCESSFUL  DAY 1     Food Truck / Cart  10017961
686327     10422594  SUCCESSFUL  DAY 1  Fast Food Restaurant  10422594

[229594 rows x 5 columns]

In [138]:
# Doing a grouping by 'merchant_category' and calculating the counting 'Successful' 'result', selecting the top 5 
# largest results of the transactions, and assigning the result in the variable 'answer3_prev2'.
# Again the result will be a data structure type of 'Series'
answer3_prev2 = answer3_prev.groupby(['merchant_category'])['result'].count().nlargest(5)

In [123]:
# Preview on 'answer3_prev2' Series. This will be illustrated with an iterative FOR loop.
answer3_prev2

merchant_category
Food / Grocery          118950
Other Retail             30454
Fast Food Restaurant     16824
Café / Restaurant        10203
Apparel                   7699
Name: result, dtype: int64

In [124]:
# Question 3 pre-answer
print('The top five Merchant Categories that made the most Successful transactions during day 1 are:')
for k in range(0,answer3_prev2.size):
    print('- ',answer3_prev2.index[k],': {:,.0f}'.format(answer3_prev2[k]))

The top five Merchant Categories that made the most Successful transactions during day 1 are:
-  Food / Grocery : 118,950
-  Other Retail : 30,454
-  Fast Food Restaurant : 16,824
-  Café / Restaurant : 10,203
-  Apparel : 7,699


The top one Merchant Category that made the most <i>Successful</i> transactions during <i>day 1</i> is:
* <i>Food / Grocery Category</i> with <i>118,950</i> successfully transactions during day 1.

=====================================================================================================================
<b>4.</b> What portion of the transactions were done using each entry mode?

        a. Based on these results and your experience as cardholder, can you guess what are the real entry modes                 behind the Entry Modes 1,2 and 3?          

In [125]:
# To answer this question will be necessary to analyze just 'transactions' DataFrame
transactions

id   date  merchant_id    card_id card_entry_mode   amount  \
0       1605032286  DAY 1      5137058  455163604          MODE 1  24000.0   
1       1605032299  DAY 1      4457960  418035837          MODE 2   4500.0   
2       1605032302  DAY 1      8669992  442446705          MODE 2   3200.0   
3       1605032254  DAY 1      4457960  418035837          MODE 1   4500.0   
4       1605032260  DAY 1      5127389  407122304          MODE 1  12000.0   
...            ...    ...          ...        ...             ...      ...   
784021  1616153350  DAY 3      6857379  328303062          MODE 2   4200.0   
784022  1616153371  DAY 3      9286781  300043014          MODE 1   2700.0   
784023  1616153397  DAY 3      5907093  239240589          MODE 3   6000.0   
784024  1616153405  DAY 3      1428282  426047641          MODE 1   7500.0   
784025  1616153378  DAY 3      4708365  426731435          MODE 1   5000.0   

            result  
0       SUCCESSFUL  
1       SUCCESSFUL  
2       SUCCESSFUL  
3           FAILED  
4       SUCCESSFUL  
...            ...  
784021      FAILED  
784022  SUCCESSFUL  
784023  SUCCESSFUL  
784024  SUCCESSFUL  
784025  SUCCESSFUL  

[781970 rows x 7 columns]

In [126]:
# Counting all the values of 'card_entry_mode' and assigning it to the variable 'total_card_entry_mode'
total_card_entry_mode = transactions['card_entry_mode'].count()

In [137]:
# Checking 'total_card_entry_mode' result
total_card_entry_mode

781970

In [184]:
# Doing a grouping by 'card_entry_mode' and calculating the counting the same column 'card_entry_mode'
# Again the result will be a data structure type of 'Series', assigned to variable 'card_entry_modes'
card_entry_modes = transactions.groupby(['card_entry_mode'])['card_entry_mode'].count()

In [185]:
# Preview on 'card_entry_modes' Series. Again, this will be illustrated with an iterative FOR loop.
card_entry_modes

card_entry_mode
MODE 1    460965
MODE 2    236635
MODE 3     84370
Name: card_entry_mode, dtype: int64

In [142]:
# Question 4 answer
print('The portion of the transactions by Entry Modes are:')
for x in range(0,card_entry_modes.size):
    print('- ',card_entry_modes.index[x],' is: {:,.2f}%'
          .format(card_entry_modes[x]/total_card_entry_mode*100),'that is {:,.0f}'.format(card_entry_modes[x]),
        'transactions of the {:,.0f}'.format(total_card_entry_mode),'transactions done.')

The portion of the transactions by Entry Modes are:
-  MODE 1  is: 58.95% that is 460,965 transactions of the 781,970 transactions done.
-  MODE 2  is: 30.26% that is 236,635 transactions of the 781,970 transactions done.
-  MODE 3  is: 10.79% that is 84,370 transactions of the 781,970 transactions done.


Based on my experience as a cardholder, I believe that the 3 types of Entry Modes refer to forms of payment available with SumUp, where I believe I identify Payments Apps (such as GooglePay or ApplePay), payment by inserting the card (with chip) and payment without contact (when the card is swiped over the SumUp device).

From the results obtained, I could assume:
* <i>MODE 1</i> corresponds to payment without contact (swiped over)
* <i>MODE 2</i> corresponds to payment by inserting the card (with chip)
* <i>MODE 3</i> corresponds to Payments Apps

=====================================================================================================================
<b>5.</b> In the PayOps team our main purpose is to make sure that our merchants receive their sales in their bank accounts in a timely manner. How many banks did we have to transfer money to, in order to pay the merchant's sales from day 2 and day 3?

In [144]:
# To solve this question I have decided to merge three DataFrames that I consider relevant for
# reply appropriately. As I did before, I will start by selecting the columns of interest, and assigning them
# to variables that will reference their content.
trans_prev5 = transactions.loc[:,['merchant_id','date','amount','result']]
merch_prev5 = merchants.loc[:,['bank_account_id', 'id']]
bank_prev5 = bank_accounts.loc[:,['bank_name', 'id']]

In [165]:
# As I did before, I'll merge 'trans_prev5' (left DataFrame) and 'merch_prev5' (right DataFrame).
# Filtering by 'Successful' transactions is mandatory according the context to answer.
# Assigning the DataFrame to get in the variable 'answer5_prev1'.
answer5_prev_1 = pd.merge(trans_prev5[trans_prev5.result == 'SUCCESSFUL'], merch_prev5, how = 'inner',
                   left_on = 'merchant_id', right_on = 'id')

In [167]:
# Preview on 'answer5_prev1' DataFrame, checking the proper columns selected and the data volume to move forward with
# this question.
answer5_prev_1

merchant_id   date   amount      result  bank_account_id        id
0           5137058  DAY 1  24000.0  SUCCESSFUL        6834904.0   5137058
1           5137058  DAY 1  16200.0  SUCCESSFUL        6834904.0   5137058
2           5137058  DAY 1   3700.0  SUCCESSFUL        6834904.0   5137058
3           5137058  DAY 1  15000.0  SUCCESSFUL        6834904.0   5137058
4           5137058  DAY 1  42000.0  SUCCESSFUL        6834904.0   5137058
...             ...    ...      ...         ...              ...       ...
746665     10432888  DAY 3   6000.0  SUCCESSFUL        9037293.0  10432888
746666     10364267  DAY 3  15000.0  SUCCESSFUL        8682059.0  10364267
746667     10609760  DAY 3   5000.0  SUCCESSFUL        9081287.0  10609760
746668     10873272  DAY 3   1000.0  SUCCESSFUL        9301512.0  10873272
746669      9586213  DAY 3  24000.0  SUCCESSFUL        9273893.0   9586213

[746670 rows x 6 columns]

In [178]:
# This second merge will allow me to join the matching values 'transactions', 'merchants' and 'bank_account',
# according to the respective match of their IDs. With this, I will know which banks to transfer for sales
# processed by merchants.
# Assigning the DataFrame to get in the variable 'answer5_prev2'.
answer5_prev_2 = pd.merge(answer5_prev_1, bank_prev5, how = 'inner',left_on = 'bank_account_id', right_on = 'id')

In [181]:
# Preview on 'answer5_prev2' DataFrame, checking the proper columns selected and the data volume to move forward with
# this question.
answer5_prev_2

merchant_id   date   amount      result  bank_account_id      id_x  \
0           5137058  DAY 1  24000.0  SUCCESSFUL        6834904.0   5137058   
1           5137058  DAY 1  16200.0  SUCCESSFUL        6834904.0   5137058   
2           5137058  DAY 1   3700.0  SUCCESSFUL        6834904.0   5137058   
3           5137058  DAY 1  15000.0  SUCCESSFUL        6834904.0   5137058   
4           5137058  DAY 1  42000.0  SUCCESSFUL        6834904.0   5137058   
...             ...    ...      ...         ...              ...       ...   
746665     10432888  DAY 3   6000.0  SUCCESSFUL        9037293.0  10432888   
746666     10364267  DAY 3  15000.0  SUCCESSFUL        8682059.0  10364267   
746667     10609760  DAY 3   5000.0  SUCCESSFUL        9081287.0  10609760   
746668     10873272  DAY 3   1000.0  SUCCESSFUL        9301512.0  10873272   
746669      9586213  DAY 3  24000.0  SUCCESSFUL        9273893.0   9586213   

       bank_name       id_y  
0         BANK 1  6834904.0  
1         BANK 1  6834904.0  
2         BANK 1  6834904.0  
3         BANK 1  6834904.0  
4         BANK 1  6834904.0  
...          ...        ...  
746665    BANK 1  9037293.0  
746666    BANK 1  8682059.0  
746667    BANK 2  9081287.0  
746668    BANK 1  9301512.0  
746669    BANK 1  9273893.0  

[746670 rows x 8 columns]

In [186]:
# Doing a grouping by 'date' and calculating the unique column of 'bank_name' to know the total banks to transfer per
# day. Again the result will be a data structure type of 'Series', assigned to variable 'banks_to_transfer_per_day'.
banks_to_transfer_per_day = answer5_prev_2.groupby('date')['bank_name'].nunique()

In [188]:
# Preview on 'banks_to_transfer_per_day' Series, checking the result to move forward with this question.
banks_to_transfer_per_day

date
DAY 1    12
DAY 2    12
DAY 3    12
Name: bank_name, dtype: int64

In [190]:
# Doing a grouping by 'date' and 'bank_name' and counting 'bank_name' to know the total transfer to do to each bank
# per day
# Again the result will be a data structure type of 'Series', assigned to variable 'num_bank_transfer_per_day'.
num_bank_transfer_per_day = answer5_prev_2.groupby(('date','bank_name'))['bank_name'].count() 

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: FutureWarning: Interpreting tuple 'by' as a list of keys, rather than a single key. Use 'by=[...]' instead of 'by=(...)'. In the future, a tuple will always mean a single key.
  after removing the cwd from sys.path.


In [191]:
# Preview on 'num_bank_transfer_per_day' Series, checking the result to move forward with this question.
# This coud be considered into the final answer, however is an multi_index series, so I'll be back to this at the end
# of the resolution of the challenge, if the time allows me.
num_bank_transfer_per_day

date   bank_name
DAY 1  BANK 1       158854
       BANK 10        1308
       BANK 11         236
       BANK 12       12887
       BANK 2        23526
       BANK 3        13494
       BANK 4        13456
       BANK 5         2654
       BANK 6         2612
       BANK 7          229
       BANK 8          236
       BANK 9          102
DAY 2  BANK 1       182922
       BANK 10        1628
       BANK 11         306
       BANK 12       14899
       BANK 2        27739
       BANK 3        15705
       BANK 4        15658
       BANK 5         3494
       BANK 6         3155
       BANK 7          247
       BANK 8          218
       BANK 9          119
DAY 3  BANK 1       179268
       BANK 10        1578
       BANK 11         120
       BANK 12       11979
       BANK 2        24370
       BANK 3        13345
       BANK 4        14523
       BANK 5         2652
       BANK 6         2761
       BANK 7          161
       BANK 8          217
       BANK 9           12
Name: bank_

In [193]:
# Doing a grouping by 'date' and counting 'bank_name' to know the total transfer to do to all banks per day
# Again the result will be a data structure type of 'Series', assigned to variable 'transfer_per_day'.
transfer_per_day = answer5_prev_2.groupby(('date'))['bank_name'].count()

In [228]:
# Preview on 'transfer_per_day' Series, checking the result to move forward with this question.
transfer_per_day

date
DAY 1    229594
DAY 2    266090
DAY 3    250986
Name: bank_name, dtype: int64

In [229]:
# Question 5 answer. The range of FOR loop starts in 1, in order to avoid the index of day 1 that is out of the scope 
# of this question.
print('The total number of banks transfers to do daily are:\n')
for y in range(1,transfer_per_day.size):
    print('-',transfer_per_day.index[y],':{:,.0f}'.format(transfer_per_day[y]),'transactions done. To',
          banks_to_transfer_per_day[y],'different banks')

The total number of banks transfers to do daily are:

- DAY 2 :266,090 transactions done. To 12 different banks
- DAY 3 :250,986 transactions done. To 12 different banks


=====================================================================================================================
<b>6.</b> Another important role you would play in the team is making sure that transactions are flowing correctly through the complex system of issuing banks, card brands, acquirers and our card readers. The KPI used to measure this is the 'Acceptance Rate', which is calculated by dividing the SUCCESFUL transactions over the TOTAL number of transactions.

        a. In the dataset you have full detail to calculate the 'Acceptance Rate' (AR) for different dimensions.   For example, you can compare the AR for the different card brands or card types and conclude over those                results. We want to suggets how we could improve the AR metric based on the data you have available.

In [231]:
# Again I will select the columns to analyze at the beginning, assigning the values of 'transactions' DataFrame in
# the 'trans_prev6'.
trans_prev6 = transactions.loc[:,['card_id','result','merchant_id']]

In [232]:
# Preview on 'trans_prev6' DataFrame, checking the proper columns selected and the data volume to move forward with
# this question.
trans_prev6

card_id      result  merchant_id
0       455163604  SUCCESSFUL      5137058
1       418035837  SUCCESSFUL      4457960
2       442446705  SUCCESSFUL      8669992
3       418035837      FAILED      4457960
4       407122304  SUCCESSFUL      5127389
...           ...         ...          ...
784021  328303062      FAILED      6857379
784022  300043014  SUCCESSFUL      9286781
784023  239240589  SUCCESSFUL      5907093
784024  426047641  SUCCESSFUL      1428282
784025  426731435  SUCCESSFUL      4708365

[781970 rows x 3 columns]

In [233]:
# As I did before, I'll merge 'trans_prev6' (left DataFrame) and 'cards' (right DataFrame).
# Assigning the DataFrame to get in the variable 'answer6_prev'.
answer6_prev = pd.merge(trans_prev6, cards, how = 'inner', left_on = 'card_id', right_on = 'id')

In [234]:
# Preview on 'answer6_prev' DataFrame, checking the proper columns selected and the data volume to move forward with
# this question.
answer6_prev

card_id      result  merchant_id         id card_type  card_brand
0       455163604  SUCCESSFUL      5137058  455163604     DEBIT        VISA
1       418035837  SUCCESSFUL      4457960  418035837     DEBIT        VISA
2       418035837      FAILED      4457960  418035837     DEBIT        VISA
3       442446705  SUCCESSFUL      8669992  442446705     DEBIT        VISA
4       407122304  SUCCESSFUL      5127389  407122304     DEBIT  MASTERCARD
...           ...         ...          ...        ...       ...         ...
781964  280062012  SUCCESSFUL      6498640  280062012     DEBIT        VISA
781965  448901621  SUCCESSFUL      8144344  448901621     DEBIT        VISA
781966  255625057  SUCCESSFUL     10114272  255625057     DEBIT        VISA
781967  328303062      FAILED      6857379  328303062     DEBIT        VISA
781968  300043014  SUCCESSFUL      9286781  300043014     DEBIT        VISA

[781969 rows x 6 columns]

In [235]:
# Doing a grouping by 'card_type' and counting 'card_type' to know the total card type involved in all transactions
# with successful and failed result included.
# Again the result will be a data structure type of 'Series', assigned to variable 'total_trans_per_card_type'.
total_trans_per_card_type = answer6_prev.groupby(('card_type'))['card_type'].count()

In [236]:
# Preview on 'total_trans_per_card_type' Series, checking the result to move forward with this question.
total_trans_per_card_type

card_type
CREDIT     27957
DEBIT     754012
Name: card_type, dtype: int64

In [237]:
# Doing a grouping by 'card_brand' and counting 'card_brand' to know the total card brand involved in all transactions
# with successful and failed result included.
# Again the result will be a data structure type of 'Series', assigned to variable 'total_trans_per_card_brand'.
total_trans_per_card_brand = answer6_prev.groupby(('card_brand'))['card_brand'].count()

In [238]:
# Preview on 'total_trans_per_card_brand' Series, checking the result to move forward with this question.
total_trans_per_card_brand

card_brand
MAESTRO       126289
MASTERCARD     65812
VISA          589868
Name: card_brand, dtype: int64

In [239]:
# Once again, I'll merge 'trans_prev6' (left DataFrame) and 'cards' (right DataFrame).
# But this time filtering by successful result, in order to complete be able to calculate the AR.
# Assigning the DataFrame to get in the variable 'answer6_prev_2'.
answer6_prev2 = pd.merge(trans_prev6[trans_prev6.result == 'SUCCESSFUL'], cards, how = 'inner',
                   left_on = 'card_id', right_on = 'id')

In [431]:
# Preview on 'answer6_prev2' DataFrame, checking the proper columns selected and the data volume to move forward with
# this question.
answer6_prev2

card_id      result  merchant_id         id card_type  card_brand
0       455163604  SUCCESSFUL      5137058  455163604     DEBIT        VISA
1       418035837  SUCCESSFUL      4457960  418035837     DEBIT        VISA
2       442446705  SUCCESSFUL      8669992  442446705     DEBIT        VISA
3       407122304  SUCCESSFUL      5127389  407122304     DEBIT  MASTERCARD
4       463103208  SUCCESSFUL     10842385  463103208     DEBIT        VISA
...           ...         ...          ...        ...       ...         ...
747517  280062012  SUCCESSFUL      6498640  280062012     DEBIT        VISA
747518  448901621  SUCCESSFUL      8144344  448901621     DEBIT        VISA
747519  255625057  SUCCESSFUL     10114272  255625057     DEBIT        VISA
747520  300043014  SUCCESSFUL      9286781  300043014     DEBIT        VISA
747521  239240589  SUCCESSFUL      5907093  239240589     DEBIT     MAESTRO

[747522 rows x 6 columns]

In [240]:
# Doing a grouping by 'card_type' and counting 'card_type' to know the total card type involved in all transactions
# Just with successful result included.
# Again the result will be a data structure type of 'Series', assigned to variable 'total_succes_trans_per_card_type'.
total_succes_trans_per_card_type = answer6_prev2.groupby(('card_type'))['card_type'].count()

In [241]:
# Preview on 'total_succes_trans_per_card_type' Series, checking the result to move forward with this question.
total_succes_trans_per_card_type

card_type
CREDIT     25339
DEBIT     722183
Name: card_type, dtype: int64

In [243]:
# Finnaly, grouping by 'card_brand' and counting 'card_brand' to know the total card type involved in all transactions
# Just with successful result included.
# Again the result will be a data structure type of 'Series', assigned to variable 'total_succes_trans_per_card_brand'
total_succes_trans_per_card_brand = answer6_prev2.groupby(('card_brand'))['card_brand'].count()

In [244]:
# Preview on 'total_succes_trans_per_card_brand' Series, checking the result to move forward with this question.
total_succes_trans_per_card_brand

card_brand
MAESTRO       119784
MASTERCARD     61866
VISA          565872
Name: card_brand, dtype: int64

In [245]:
# Calculation of AR by Cards Types
for z in range(0,total_succes_trans_per_card_type.size):
    print('The AR by',total_succes_trans_per_card_type.index[z],'Card type is: {:,.3f}'
          .format(total_succes_trans_per_card_type[z]/total_trans_per_card_type[z]),
          '. Base on {:,.0f}'.format(total_trans_per_card_type[z]),'total transactions of',
          total_succes_trans_per_card_type.index[z],'cards, and {:,.0f}'.format(total_succes_trans_per_card_type[z]),
          'Successfull transactions of the same card type.\n')

The AR by CREDIT Card type is: 0.906 . Base on 27,957 total transactions of CREDIT cards, and 25,339 Successfull transactions of the same card type.

The AR by DEBIT Card type is: 0.958 . Base on 754,012 total transactions of DEBIT cards, and 722,183 Successfull transactions of the same card type.



In [246]:
# Calculation of AR by Cards Brands
for i in range(0,total_succes_trans_per_card_brand.size):
    print('The AR by',total_succes_trans_per_card_brand.index[i],'Card brand is: {:,.3f}'
          .format(total_succes_trans_per_card_brand[i]/total_trans_per_card_brand[i]),
          '. Base on {:,.0f}'.format(total_trans_per_card_brand[i]),'total transactions of',
          total_succes_trans_per_card_brand.index[i],
          'cards, and {:,.0f}'.format(total_succes_trans_per_card_brand[i]),
          'Successfull transactions of the same card brand.\n')

The AR by MAESTRO Card brand is: 0.948 . Base on 126,289 total transactions of MAESTRO cards, and 119,784 Successfull transactions of the same card brand.

The AR by MASTERCARD Card brand is: 0.940 . Base on 65,812 total transactions of MASTERCARD cards, and 61,866 Successfull transactions of the same card brand.

The AR by VISA Card brand is: 0.959 . Base on 589,868 total transactions of VISA cards, and 565,872 Successfull transactions of the same card brand.



Regarding how to improve the AR KPI, according to what was calculated and the data available in this data set, I think it would be interesting to be able to cluster it in a better way, like using merchants categories, as well as by individual merchants and even by comunas.

Perhaps there are particular behaviors in the payments of clients that use a certain entry_card_mode, for a certain business segment in a certain comunas, with a certain card type and card brand.

In this way we would be able to know in what type of businesses, or in which comunas there are greater failures in the payment processes and what type of entry_card_mode, card type or brand are related. This may allow us to take certain types of measures and become aware of external factors, maybe the SumUp device are being used in a merchant located in a place with such as a failure in the connection to the network or problems with the charging of device, or in a place with temperature and/or humidity conditions that may affect the correct functioning of the devices if we are able to determine that the greatest failures in the payment processes occur in places that meet these characteristics.

Another benefit of segmenting by comunas is the chance to geolocatize all the SumUp clients and their respective devices, which could be compared with the client delivery date of the device and see if it is a problem in which a scheduled replacement of the device could mitigate and in this way program SumUp device replacements between the different devices distributed throughout the city's comunas, prioritizing those that are further away and generate high transaction traffic with a medium or low AR, to mention some ideas.

In [445]:
# Preview of answer6_prev

In [446]:
answer6_prev

card_id      result  merchant_id         id card_type  card_brand
0       455163604  SUCCESSFUL      5137058  455163604     DEBIT        VISA
1       418035837  SUCCESSFUL      4457960  418035837     DEBIT        VISA
2       418035837      FAILED      4457960  418035837     DEBIT        VISA
3       442446705  SUCCESSFUL      8669992  442446705     DEBIT        VISA
4       407122304  SUCCESSFUL      5127389  407122304     DEBIT  MASTERCARD
...           ...         ...          ...        ...       ...         ...
781964  280062012  SUCCESSFUL      6498640  280062012     DEBIT        VISA
781965  448901621  SUCCESSFUL      8144344  448901621     DEBIT        VISA
781966  255625057  SUCCESSFUL     10114272  255625057     DEBIT        VISA
781967  328303062      FAILED      6857379  328303062     DEBIT        VISA
781968  300043014  SUCCESSFUL      9286781  300043014     DEBIT        VISA

[781969 rows x 6 columns]

In [248]:
# one more merge with merchants
answer7_prev = pd.merge(answer6_prev, merchants, how = 'inner',
                   left_on = 'merchant_id', right_on = 'id')


In [249]:
# gruping and counting one more time for all transactions
merch_trans = answer7_prev.groupby(('merchant_category'))['merchant_category'].count()

In [255]:
# and for successful transactions
answer6_prev3 = pd.merge(answer6_prev[answer6_prev.result == 'SUCCESSFUL'], merchants, how = 'inner',
                   left_on = 'merchant_id', right_on = 'id')

In [256]:
merch_succes_trans = answer6_prev3.groupby(('merchant_category'))['merchant_category'].count()

In [257]:
# Checking AR by Merchants Categories
for h in range(0,merch_succes_trans.size):
    print('The AR by',merch_succes_trans.index[h],'merchant Category is: {:,.3f}'
          .format(merch_succes_trans[h]/merch_trans[h]),
          '. Base on {:,.0f}'.format(merch_trans[h]),'total transactions of',
          merch_succes_trans.index[h],
          'merchant, and {:,.0f}'.format(merch_succes_trans[h]),
          'Successfull transactions of the same merchant category.\n')

The AR by Apparel merchant Category is: 0.951 . Base on 26,813 total transactions of Apparel merchant, and 25,493 Successfull transactions of the same merchant category.

The AR by Art Dealers and Galleries merchant Category is: 0.964 . Base on 2,863 total transactions of Art Dealers and Galleries merchant, and 2,761 Successfull transactions of the same merchant category.

The AR by Bar / Club merchant Category is: 0.931 . Base on 5,953 total transactions of Bar / Club merchant, and 5,543 Successfull transactions of the same merchant category.

The AR by Beauty / Barber merchant Category is: 0.956 . Base on 13,233 total transactions of Beauty / Barber merchant, and 12,650 Successfull transactions of the same merchant category.

The AR by Café / Restaurant merchant Category is: 0.959 . Base on 29,046 total transactions of Café / Restaurant merchant, and 27,863 Successfull transactions of the same merchant category.

The AR by Car Repair Shop merchant Category is: 0.945 . Base on 4,503 t

In [258]:
# Checking AR by Merchants Categories, and maybe putting some alerts if is required to alert if some KPI has a 
# negative variation.
for n in range(0,merch_succes_trans.size):
    if((merch_succes_trans[n]/merch_trans[n] > 0.92) & (merch_succes_trans[n]/merch_trans[n] < 0.96)):
        statement = 'Moderate AR. Could be better'
    elif (merch_succes_trans[n]/merch_trans[n] >= 0.96):
        statement = 'Good KPI!!!'
    else:
        statement = 'Review ASAP!!!'
    
    print('The AR by',merch_succes_trans.index[h],'merchant Category is: {:,.3f}'
          .format(merch_succes_trans[n]/merch_trans[n]),
          '. Base on {:,.0f}'.format(merch_trans[n]),'total transactions of',
          merch_succes_trans.index[n],
          'merchant, and {:,.0f}'.format(merch_succes_trans[n]),
          'Successfull transactions of the same merchant category.',statement,'\n\n')

The AR by Veterinary merchant Category is: 0.951 . Base on 26,813 total transactions of Apparel merchant, and 25,493 Successfull transactions of the same merchant category. Moderate AR. Could be better 


The AR by Veterinary merchant Category is: 0.964 . Base on 2,863 total transactions of Art Dealers and Galleries merchant, and 2,761 Successfull transactions of the same merchant category. Good KPI!!! 


The AR by Veterinary merchant Category is: 0.931 . Base on 5,953 total transactions of Bar / Club merchant, and 5,543 Successfull transactions of the same merchant category. Moderate AR. Could be better 


The AR by Veterinary merchant Category is: 0.956 . Base on 13,233 total transactions of Beauty / Barber merchant, and 12,650 Successfull transactions of the same merchant category. Moderate AR. Could be better 


The AR by Veterinary merchant Category is: 0.959 . Base on 29,046 total transactions of Café / Restaurant merchant, and 27,863 Successfull transactions of the same merchant

=====================================================================================================================
<b>7.</b> Bonus: share 1 insight that you think is relevant and interesting for us to know and base it on a new SQL query different to the ones ued before.

Maybe we can focus on improving the Acceptance Rate, and on understanding where and how the biggest payment failures occur. A bit following the above logic. As for what type of card entry mode the highest number of failures occur, segmented by merchant category and district.

Perhaps the failures are periodic problems, for example, they always occur at the end of the month or they may be correlated with a particular event that occurs from time to time, for which it will be important to have the date of the failed transaction.

Based on the shared data set, and an entity-relationship diagram draft that I prepared to better illustrate myself in the development of this test, I could answer this through the following SQL query:

<img src="ER_Draft.jpg">

WITH sb1 AS 
(SELECT t.date, t.merchant_id, t.card_entry_mode, t.amount, c.card_type, c.card_brand
FROM transactions AS t
INNER JOIN cards AS c
ON t.card_id = c.id
WHERE result = 'FAILED'
)

SELECT sb1.card_entry_mode, sb1.card_type, sb1.card_brand, m.merchant_category, m.comuna_id, sb1.date, sb1.amount
FROM sb1
INNER JOIN merchants AS m
ON sb1.merchant_id = m.id
GROUP BY m.comuna_id, m.merchant_category, sb1.card_entry_mode, sb1.card_type, sb1.card_brand;